### PR TITLE
feat(devops): Add Init Args

### PR DESCRIPTION
**Pull request by lmuntaner**
# Motivation

To validate credentials Oisy backend will need some information about the environment. Such as the canister ids of II and the POUH issuer.

In this PR, I add new fields to InitArg and pass them on deploy.

# Changes

* Add fields to InitArg.
* Gather the environment information in deploy.backend and pass it as init arguments.
* Update candid file with new fields and type.
* Update the config on post_upgrade and not only on init.

# Tests

I tested by deploying locally the backend to an already deployed canister and it worked.

### CODE REVIEW

File: scripts/deploy.backend.sh
Comment (peterpeterparker) [inline scripts/deploy.backend.sh:17]: This is not incorrect for a deployment on mainnet (production and staging)?

File: scripts/deploy.backend.sh
Comment (peterpeterparker) [inline scripts/deploy.backend.sh:18]: ditto

File: scripts/deploy.backend.sh
Comment (peterpeterparker) [inline scripts/deploy.backend.sh:21]: The contrary, that's invalid for local and ci development no?

File: scripts/deploy.backend.sh
Comment (peterpeterparker) [inline scripts/deploy.backend.sh:28]: For readability reason, I would move the initialization of any variable depending of the environment in the existing `case` block?

File: src/backend/src/lib.rs
Comment (peterpeterparker) [inline src/backend/src/lib.rs:160]: Why do we need this root key in the state? Can't it be derived on the fly when needed? If it has to be saved in the state, what are the security implication?

File: src/backend/src/lib.rs
Comment (peterpeterparker) [inline src/backend/src/lib.rs:170]: It's an acceptable error - i.e. if it fails it's ok to let the init and upgrade of the canister succeed? How do we notice if this fails?

File: src/backend/src/lib.rs
Comment (peterpeterparker) [inline src/backend/src/lib.rs:169]: There is no enforcement in place to ensure that no root_key args is passed when the canister is deployed on mainnet, is that safe enough?

File: src/backend/src/lib.rs
Comment (peterpeterparker) [inline src/backend/src/lib.rs:179]: Not sure but you might be able to destruct the arg here to be less version.

```
Config {
   ic_root_key_raw,
   ..arg
}
```

But agin, not sure.

File: src/backend/src/lib.rs
Comment (peterpeterparker) [inline src/backend/src/lib.rs:163]: Given that the function does not replace a single fiel but entirely overwrite the config, maybe `init_config` instead of `update_config`?

File: src/backend/src/lib.rs
Comment (peterpeterparker) [inline src/backend/src/lib.rs:207]: `None` is acceptable? If yes you can maybe spare some code duplication by replacing `Some(Arg::Upgrade)` and `None` with `_`.

If `None` is not acceptable, maybe trap?

File: src/shared/src/types.rs
Comment (peterpeterparker) [inline src/shared/src/types.rs:7]: What happens if we support both `ic0.app` and `internetcomputer.org`? Not an issue with a single `ii_origin`?

File: scripts/deploy.backend.sh
Comment (lmuntaner) [reply scripts/deploy.backend.sh:17]: Why would this be incorrect for production and staging?

File: scripts/deploy.backend.sh
Comment (lmuntaner) [reply scripts/deploy.backend.sh:18]: Also not, the proper canister id will be set in dfx.json in this case.

File: scripts/deploy.backend.sh
Comment (lmuntaner) [reply scripts/deploy.backend.sh:21]: No, that's not correct. Maybe I'll rename the variable, but in the library it's using this name.

That's why I added a comment about, but this is more of an ID than a URL. The credential will have this url hardcoded and we need to compare it with this.

File: scripts/deploy.backend.sh
Comment (lmuntaner) [reply scripts/deploy.backend.sh:28]: Yes, good point.

File: src/backend/src/lib.rs
Comment (lmuntaner) [reply src/backend/src/lib.rs:160]: The root key in a local replica changes every time. We already discussed with the crypto team to have a fixed one, but that won't happen in the short term.

Moreover, the root key is not accessible from within a canister.

Therefore, the root key needs to be passed as a parameter on init (or upgrade).

This is how we have been keeping the root key in all the demos we have been doing and the feedback from security was to have a hardcoded mainnet key.

File: src/backend/src/lib.rs
Comment (lmuntaner) [reply src/backend/src/lib.rs:163]: Good point, how about `set_config` then. Init seems to be done only once.

File: src/backend/src/lib.rs
Comment (lmuntaner) [reply src/backend/src/lib.rs:169]: I believe so, but I wonder whether we should ask for a security review after this PR?

File: src/backend/src/lib.rs
Comment (lmuntaner) [reply src/backend/src/lib.rs:170]: Mhm, I don't know. My default is always to keep running code. But maybe in this case we want to panic and skip the upgrade. I'll add a `panic` instead.

File: src/backend/src/lib.rs
Comment (lmuntaner) [reply src/backend/src/lib.rs:179]: Sure!

File: src/backend/src/lib.rs
Comment (lmuntaner) [reply src/backend/src/lib.rs:207]: I believe it's ok yes. One could upgrade without an argument and we still want to check the config.

I'll use `_` thanks!

File: src/shared/src/types.rs
Comment (lmuntaner) [reply src/shared/src/types.rs:7]: Not an issue, the credentials from both will have the same value.

I believe this is again a naming confusion.

As an example you can try to the VC flow with the dummy [relying party](https://l7rua-raaaa-aaaap-ahh6a-cai.icp0.io/) and dummy issuer (default values).

After receiving the credentials, you can decode them and then decode the alias credential.

The "iss" field is what this `ii_origin` means. It's the issuer.

Screenshot 2024-07-16 at 13 13 48

I used this names (issuer_origin also) because it's the name used i[n the library](https://github.com/dfinity/verifiable-credentials-sdk/blob/a116a2e53a13e4a1f6f7e8429960a7d0d6142e70/rust-packages/ic-verifiable-credentials/src/lib.rs#L45) that Bartosz created to validate the credentials.

I'm ok changing the name here and then mapping it properly. I also found it confusing from the coding perspective, but it made sense from the standard perspective.

File: scripts/deploy.backend.sh
Comment (peterpeterparker) [reply scripts/deploy.backend.sh:33]: Right, it works for production. Doesn't for staging though

> ❯ dfx canister id internet_identity --network staging
> Error: Cannot find canister id. Please issue 'dfx canister create internet_identity --network staging'.

File: scripts/deploy.backend.sh
Comment (peterpeterparker) [reply scripts/deploy.backend.sh:34]: Right.

File: scripts/deploy.backend.sh
Comment (peterpeterparker) [reply scripts/deploy.backend.sh:37]: Oh it's an ID. Yeah would be nice to rename if possible. Not necessary in this PR but, generally speaking.

File: scripts/deploy.backend.sh
Comment (peterpeterparker) [inline scripts/deploy.backend.sh:50]: Seeing it now, it's always "ProofOfUniqueness" - i.e. shouldn't the type be a variant / enum instead of String?

File: src/backend/src/lib.rs
Comment (peterpeterparker) [reply src/backend/src/lib.rs:160]: Thanks for the exaplantion, did not knew. It feels a bit weird no?
All good.

File: src/backend/src/lib.rs
Comment (peterpeterparker) [inline src/backend/src/lib.rs:171]: With the help of ChatGPT, if interesting, the two lines version of this block:

```
let ic_root_key_raw = extract_raw_root_pk_from_der(
    &ic_root_key_der.unwrap_or_else(|| IC_ROOT_PK_DER.to_vec())
).map(Some).unwrap_or_else(|_| panic!("There was an error parsing the root key"));
```

File: src/backend/tests/it/utils/pocketic.rs
Comment (peterpeterparker) [inline src/backend/tests/it/utils/pocketic.rs:95]: You provide a test for the upgrade in another PR?

File: src/shared/src/types.rs
Comment (peterpeterparker) [reply src/shared/src/types.rs:7]: I'm not sure I still get it but, thanks for the explanation. I think you are right we should use the naming convetion even if they do feel a bit weird at first sight.

That said, I'm still unsure what happens when a credential is issued on iss ic0.app but used on internetcomputer.org but, definitely nothin to do with this PR.

File: scripts/deploy.backend.sh
Comment (lmuntaner) [reply scripts/deploy.backend.sh:50]: Can I create an enum with a set of strings in Candid or Rust?

I would need the mapper of an enum / variant to a string at some point anyway.

That would mean that we'll need to add the types to the enum when we want to support new credentials. Which kind of defeats the purpose to have it as a parameter as well, doesn't it?

File: src/backend/src/lib.rs
Comment (lmuntaner) [reply src/backend/src/lib.rs:171]: 🎉

File: src/backend/tests/it/utils/pocketic.rs
Comment (lmuntaner) [reply src/backend/tests/it/utils/pocketic.rs:95]: What do you mean a test for the upgrade?

Are there already upgrade tests?

File: src/shared/src/types.rs
Comment (lmuntaner) [reply src/shared/src/types.rs:7]: internetcomputer.org will issue the same credential on ic0.app because both have the identity.ic0.app hardcoded.

File: scripts/deploy.backend.sh
Comment (lmuntaner) [reply scripts/deploy.backend.sh:37]: As discussed in [this comment](https://github.com/dfinity/oisy-wallet/pull/1773#discussion_r1679239909) we chose to keep the name according to the convention.

File: src/backend/src/lib.rs
Comment (peterpeterparker) [reply src/backend/src/lib.rs]: Maybe. Generally speaking we have been keeping them in the loop of Oisy each time we have build a new feature.

File: scripts/deploy.backend.sh
Comment (peterpeterparker) [reply scripts/deploy.backend.sh:50]: I think so. In Rust you create an enum which translate to a variant in Candid.

For example this enum: https://github.com/junobuild/juno/blob/e41584c130736ddea8c8e2bd0a444119136599d4/src/libs/shared/src/types.rs#L32C5-L36C6

Becomes this variant in Candid: https://github.com/junobuild/juno/blob/e41584c130736ddea8c8e2bd0a444119136599d4/src/declarations/observatory/observatory.did.d.ts#L10

So I guess here we can have 

```
 #[derive(CandidType, Serialize, Deserialize, Clone)]
pub enum CredentialType {
       ProofOfUniqueness,
}
```

But, I have no clue of the credentials mechanism, just a suggestion if the type are defined in some ways.

File: src/backend/tests/it/utils/pocketic.rs
Comment (peterpeterparker) [reply src/backend/tests/it/utils/pocketic.rs:95]: Sure, for tokens when we added the flag `version` and `enabled`. There are also utils to dowload a particular wasm of the backend and then proceed with an upgrade.

See https://github.com/dfinity/oisy-wallet/tree/main/src/backend/tests/it/upgrade

File: scripts/deploy.backend.sh
Comment (lmuntaner) [reply scripts/deploy.backend.sh:50]: The credential type in Rust is a String:

https://github.com/dfinity/verifiable-credentials-sdk/blob/a116a2e53a13e4a1f6f7e8429960a7d0d6142e70/rust-packages/ic-verifiable-credentials/src/issuer_api.rs#L94

File: src/backend/src/lib.rs
Comment (lmuntaner) [reply src/backend/src/lib.rs]: I'll do that then!

Comment (lmuntaner) [reply src/backend/src/lib.rs]: I'll also point to this PR as well.

**Review by peterpeterparker — APPROVED** [review]
LGTM, thanks

Might be worth to have a quick review from @bitdivine too particularly as we are touching init and post upgrade here.

File: scripts/deploy.backend.sh
Comment (bitdivine) [inline scripts/deploy.backend.sh:19]: It _seems_ to work with dfx 0.20.0, as defined in dfx.json, for mainnet and localhost:
```
$ dfx ping ic | jq -r '.root_key | reduce .[] as $item ("opt vec { "; "\(.) \($item):nat8;") + " }"'
opt vec {  48:nat8; 129:nat8; 130:nat8; 48:nat8; 29:nat8; 6:nat8; 13:nat8; 43:nat8; 6:nat8; 1:nat8; 4:nat8; 1:nat8; 130:nat8; 220:nat8; 124:nat8; 5:nat8; 3:nat8; 1:nat8; 2:nat8; 1:nat8; 6:nat8; 12:nat8; 43:nat8; 6:nat8; 1:nat8; 4:nat8; 1:nat8; 130:nat8; 220:nat8; 124:nat8; 5:nat8; 3:nat8; 2:nat8; 1:nat8; 3:nat8; 97:nat8; 0:nat8; 129:nat8; 76:nat8; 14:nat8; 110:nat8; 199:nat8; 31:nat8; 171:nat8; 88:nat8; 59:nat8; 8:nat8; 189:nat8; 129:nat8; 55:nat8; 60:nat8; 37:nat8; 92:nat8; 60:nat8; 55:nat8; 27:nat8; 46:nat8; 132:nat8; 134:nat8; 60:nat8; 152:nat8; 164:nat8; 241:nat8; 224:nat8; 139:nat8; 116:nat8; 35:nat8; 93:nat8; 20:nat8; 251:nat8; 93:nat8; 156:nat8; 12:nat8; 213:nat8; 70:nat8; 217:nat8; 104:nat8; 95:nat8; 145:nat8; 58:nat8; 12:nat8; 11:nat8; 44:nat8; 197:nat8; 52:nat8; 21:nat8; 131:nat8; 191:nat8; 75:nat8; 67:nat8; 146:nat8; 228:nat8; 103:nat8; 219:nat8; 150:nat8; 214:nat8; 91:nat8; 155:nat8; 180:nat8; 203:nat8; 113:nat8; 113:nat8; 18:nat8; 248:nat8; 71:nat8; 46:nat8; 13:nat8; 90:nat8; 77:nat8; 20:nat8; 80:nat8; 95:nat8; 253:nat8; 116:nat8; 132:nat8; 176:nat8; 18:nat8; 145:nat8; 9:nat8; 28:nat8; 95:nat8; 135:nat8; 185:nat8; 136:nat8; 131:nat8; 70:nat8; 63:nat8; 152:nat8; 9:nat8; 26:nat8; 11:nat8; 170:nat8; 174:nat8; }


$ dfx --version
dfx 0.20.0
```
But if you are getting errors in some cases, for sure, dfx needs to get fixed! :-)

File: src/backend/src/lib.rs
Comment (bitdivine) [inline src/backend/src/lib.rs:175]: So the idea is that if an incorrect argument is passed, the canister dies in the post-upgrade hook rather than proceeding?

File: src/backend/src/lib.rs
Comment (bitdivine) [inline src/backend/src/lib.rs:185]: If I read correctly, `.set()` can fail only if the data is too big to fit into memory.  The memory is expanded if necessary, so it tentatively looks to me as if the expect could be `.unwrap_or_else(|_| unreachable!("This data structure is too small to exhaust stable memory."))`?  I guess the rest of memory could be filled with accounts, and this is taking just the last few bytes so then I guess it could happen.  OK!  Please shout if I missed something but at the moment `expect` looks correct to me.

**Review by bitdivine — APPROVED** [review]
LGTM, thanks!

File: src/declarations/backend/backend.did
Comment (lmuntaner) [reply src/declarations/backend/backend.did:2]: This was tested manually and I intend to add an upgrade test in a following PR.

File: scripts/deploy.backend.sh
Comment (lmuntaner) [reply scripts/deploy.backend.sh:19]: Thanks! I changed it. I copied it from another project.

File: src/backend/src/lib.rs
Comment (lmuntaner) [reply src/backend/src/lib.rs:175]: That is correct.

File: src/backend/src/lib.rs
Comment (lmuntaner) [reply src/backend/src/lib.rs:185]: Better be on the safe side.

**Pull request merged**
